### PR TITLE
Ufe bbox support dev

### PR DESCRIPTION
**Pull request by ppt-adsk**

### CODE REVIEW

File: lib/ufe/Global.cpp [hunk 1]
```diff
@@ -93,14 +94,13 @@ MStatus initialize()
...
-		kUSDRunTimeName, usdHierHandler, usdTrans3dHandler, usdSceneItemOpsHandler, usdAttributesHandler, nullptr);
-#else
+	UFE_V2(auto usdAttributesHandler = UsdAttributesHandler::create();)
+	UFE_V2(auto usdObject3dHandler = UsdObject3dHandler::create();)
 	g_USDRtid = Ufe::RunTimeMgr::instance().register_(
-		kUSDRunTimeName, usdHierHandler, usdTrans3dHandler, usdSceneItemOpsHandler);
-#endif
+		kUSDRunTimeName, usdHierHandler, usdTrans3dHandler, 
+        usdSceneItemOpsHandler
+        UFE_V2(, usdAttributesHandler, usdObject3dHandler));
```
Comment (ppt-adsk) [inline lib/ufe/Global.cpp]: Register new implementation of UFE Object3d interface.  For the moment, this only has support for bounding box computation, which is used by Maya's viewFit command.

File: lib/ufe/UsdObject3d.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    //
+    // UsdGeomImageable::ComputeLocalBound() just calls UsdGeomBBoxCache, so do
+    // this here as well.
+    //
+    // Would be nice to know if the object extents are animated or not, so
+    // we can bypass time computation and simply use UsdTimeCode::Default()
+    // as the time.
+    TfTokenVector purposes{UsdGeomTokens->default_};
+    UsdGeomBBoxCache bboxCache(getTime(sceneItem()->path()), purposes);
+    auto bbox = bboxCache.ComputeLocalBound(fPrim);
```
Comment (ppt-adsk) [inline lib/ufe/UsdObject3d.cpp]: Use USD to compute the object's local bounding box.

File: lib/ufe/UsdObject3d.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#include "ufe/object3d.h"
+
+MAYAUSD_NS_DEF {
+namespace ufe {
+
+//! \brief USD run-time 3D object interface
+/*!
+    This class implements the Object3d interface for USD prims.
+*/
+class MAYAUSD_CORE_PUBLIC UsdObject3d : public Ufe::Object3d
```
Comment (ppt-adsk) [inline lib/ufe/UsdObject3d.h]: Implement UFE Object3d interface for USD data model.

File: lib/ufe/UsdObject3dHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+UsdObject3dHandler::Ptr UsdObject3dHandler::create()
+{
+	return std::make_shared<UsdObject3dHandler>();
+}
+
+//------------------------------------------------------------------------------
+// UsdObject3dHandler overrides
+//------------------------------------------------------------------------------
+
+Ufe::Object3d::Ptr UsdObject3dHandler::object3d(const Ufe::SceneItem::Ptr& item) const
```
Comment (ppt-adsk) [inline lib/ufe/UsdObject3dHandler.cpp]: Implement Object3d interface factory for USD.

File: lib/ufe/Utils.cpp [hunk 5]
```diff
@@ -218,5 +221,45 @@ MDagPath nameToDagPath(const std::string& name)
...
+        return UsdTimeCode::Default();
+    }
+
+    // Get the time from the proxy shape.  This will be the tail component of
+    // the first path segment.
+    auto proxyShapePath = Ufe::Path(path.getSegments()[0]);
+
+    // Keep a single-element path to MObject cache, as all USD prims in a stage
+    // share the same proxy shape.
+    static std::pair<Ufe::Path, MObjectHandle> cache;
```
Comment (ppt-adsk) [inline lib/ufe/Utils.cpp]: Implement a quickie proxy shape cache for time computation.

File: test/lib/ufe/testAttribute.py [hunk 6]
```diff
@@ -17,6 +17,7 @@
 #
 
 from ufeTestUtils import usdUtils, mayaUtils
+from ufeTestUtils.testUtils import assertVectorAlmostEqual
```
Comment (ppt-adsk) [inline test/lib/ufe/testAttribute.py]: When adding the Object3d interface test, had to use assertVectorAlmostEqual, which cause me to go beyond the copy-paste threshold of pain, so I factored it out into a utility, and deployed the utility in all tests that use assertVectorAlmostEqual.

**Comment by kxl-adsk** [comment]
@pilarmolinalopez can you help review this change?

**Comment by ppt-adsk** [comment]
Thanks Pilar for the code review, very appreciated.

File: test/lib/ufe/testAttribute.py (hunk 6 above)
Comment (seando-adsk) [reply test/lib/ufe/testAttribute.py]: Great thanks.

**Pull request merged**